src/debug_info.rs

```diff
@@ -1,30 +1,1221 @@
+    placeholders: Vec<LLVMMetadataRef>,
+    has_cu: bool,
+    _marker: PhantomData<&'ctx Context>,
+}
+
+/// Any kind of debug information scope (i.e. visibility of a source code symbol). Scopes are
+/// created by special `DebugInfoBuilder` methods (eg `create_lexical_block`) and can be turned
+/// into a `DIScope` with the `AsDIScope::as_debug_info_scope` trait method.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct DIScope<'ctx> {
+    metadata_ref: LLVMMetadataRef,
+    _marker: PhantomData<&'ctx Context>,
 }
 
-#[derive(Debug)]
-pub struct DebugInfoBuilder(LLVMDIBuilderRef);
+/// Specific scopes (i.e. `DILexicalBlock`) can be turned into a `DIScope` with the
+/// `AsDIScope::as_debug_info_scope` trait method.
+pub trait AsDIScope<'ctx> {
+    fn as_debug_info_scope(self) -> DIScope<'ctx>;
+}
+
+impl<'ctx> DebugInfoBuilder<'ctx> {
+    pub(crate) fn new(module: &Module, allow_unresolved: bool) -> Self {
+        let builder = unsafe {
+            if allow_unresolved {
+                LLVMCreateDIBuilder(module.module.get())
+            } else {
+                LLVMCreateDIBuilderDisallowUnresolved(module.module.get())
+            }
+        };
+
+        DebugInfoBuilder {
+            builder,
+            placeholders: vec![],
+            has_cu: false,
+            _marker: PhantomData,
+        }
+    }
+
+    /// A DICompileUnit provides an anchor for all debugging information generated during this instance of compilation.
+    ///
+    /// * `language` - Source programming language
+    /// * `file` - File info
+    /// * `producer` - Identify the producer of debugging information and code. Usually this is a compiler version string.
+    /// * `is_optimized` - A boolean flag which indicates whether optimization is enabled or not.
+    /// * `flags` - This string lists command line options. This string is directly embedded in debug info output which may be used by a tool analyzing generated debugging information.
+    /// * `runtime_ver` - This indicates runtime version for languages like Objective-C.
+    /// * `split_name` - The name of the file that we'll split debug info out into.
+    /// * `kind` - The kind of debug information to generate.
+    /// * `dwo_id` - The DWOId if this is a split skeleton compile unit.
+    /// * `split_debug_inlining` - Whether to emit inline debug info.
+    /// * `debug_info_for_profiling` - Whether to emit extra debug info for profile collection.
+    pub fn create_compile_unit(
+        &mut self,
+        language: DWARFSourceLanguage,
+        file: DIFile<'ctx>,
+        producer: &str,
+        is_optimized: bool,
+        flags: &str,
+        runtime_ver: libc::c_uint,
+        split_name: &str,
+        kind: DWARFEmissionKind,
+        dwo_id: libc::c_uint,
+        split_debug_inlining: bool,
+        debug_info_for_profiling: bool,
+    ) -> DICompileUnit<'ctx> {
+        if self.has_cu {
+            panic!("Only one compile unit per DebugInfoBuilder.");
```

Comment:
Does this need to panic? Shouldn't we return a Result?

Reply:
This panic! is removed.